task/download-sbom-from-url-in-attestation/0.1/download-sbom-from-url-in-attestation.yaml

```diff
@@ -0,0 +1,256 @@
+
+              auth_key=$new_key
+          done
+
+      }
+
+      download_blob() {
+        local blob_ref=$1
+        local dest=${2:-/dev/stdout}
+
+        # convert:
+        #     registry.com/namespace/repo@sha256:digest
+        # ->  https://registry.com/v2/namespace/repo/blobs/sha256:digest
+        blob_url=$(sed -E 's;([^/]*)/(.*)@(.*);https://\1/v2/\2/blobs/\3;' <<< "$blob_ref")
+
+        local tmp_dest=$(mktemp --tmpdir)
+
+        local common_curl_opts=(--silent --show-error --retry "${HTTP_RETRIES:-3}")
+
+        echo "GET $blob_url" >&2
+        local outputs
+        mapfile -t outputs < <(curl \
+            "${common_curl_opts[@]}" \
+            -L \
+            --write-out '%header{www-authenticate}\n%{response_code}' \
+            --output "$tmp_dest" \
+            "$blob_url"
+        )
+        local www_authenticate=${outputs[0]}
+        local response_code=${outputs[1]}
+
+        if [[ "$response_code" -eq 200 ]]; then
+            # Blob download didn't require auth, we're done
+            :
+        elif [[ "$response_code" -eq 401 ]]; then
+            echo "Got 401, trying to authenticate" >&2
+
+            local realm service scope token_url
+            realm=$(get_from_www_auth_header "$www_authenticate" realm)
+            service=$(get_from_www_auth_header "$www_authenticate" service)
+            scope=$(get_from_www_auth_header "$www_authenticate" scope)
+            token_url=$(jq -n -r --arg realm "$realm" --arg service "$service" --arg scope "$scope" \
+                '"\($realm)?service=\($service | @uri)&scope=\($scope | @uri)"'
+            )
+
+            local basic_auth token_auth
+            if basic_auth=$(get_container_auth "$blob_ref"); then
+                token_auth=(-H "authorization: Basic $basic_auth")
+            else
+                echo "Trying to get token anonymously" >&2
+                token_auth=()
+            fi
+
+            echo "GET $token_url" >&2
+            token=$(curl \
+                "${common_curl_opts[@]}" \
+                "${token_auth[@]}" \
+                --fail-with-body \
+                "$token_url" | jq -r .token
+            )
+
+            echo "GET $blob_url" >&2
+            curl \
+                "${common_curl_opts[@]}" \
+                -L \
+                --output "$tmp_dest" \
+                --fail-with-body \
+                -H "authorization: Bearer $token" \
+                "$blob_url"
+        else
+            echo "Error: unexpected response code: $response_code!" >&2
+            return 1
+        fi
+
+        cp "$tmp_dest" "$dest"
+      }
+
+      for image in "$@"; do
+        attestation_file="$WORKDIR/$image/attestation.json"
+        sbom_blob_url=$(jq -r \
+          '.payload | @base64d | fromjson | .. | select(.name? == "SBOM_BLOB_URL").value' \
+          "$attestation_file"
+        )
+        mkdir -p "$SBOMS_DIR/$image"
+        download_blob "$sbom_blob_url" "$SBOMS_DIR/$image/sbom.json"
```

Comment:
Alternatively, we could just get the SBOM with `cosign download sbom`. But that's even less secure than what we're doing now.

Given a container image:
* With the "find SBOM_BLOB_URL in the attestation" solution:
  * The pipeline can verify that the SBOM_BLOB_URL was returned by the Tekton pipeline that built the image. You still have to trust that the build pipeline created a real SBOM and returned the correct url.
* With the cosign solution:
  * The pipeline takes the image digest and downloads the `sha256-$digest.sbom` tag from the image repository. You have to trust (everything above and) that nobody tampered with the tag between the build and the gitops promotion.